#This script was used to prep data files into MMUPHin friendly formats. 
#MMUPHin requires two input files: feature counts table and metadata file.
#Feature counts table is generated by combining the taxonomy table and ASV/OTU table then saved as a matrix.The taxa column must be annotated in the format used below (all taxa levels combined) for MMUPHin to run effectively.

#!/usr/bin/Rscript4.2.1

library(data.table)
library(readr)
library(magrittr)
library(dplyr)
library(ggplot2)
library(tidyverse)
library(phyloseq)
library(Biostrings)
library(hiReadsProcessor)
library(ShortRead)
library(metagMisc)
library(stringr)
library(microbiome)
library(knitr)
library(DECIPHER)
library(vegan)
library(dada2)
library(phangorn)

#Specify path where FASTQ files are located
path <- "~/16S_mouse_data/MMUPHIN_SeqTabs"

#Load files
OTU <- readRDS("~/16S_mouse_data/MMUPHIN_SeqTabs/OTU_MMUPHin.rds")
taxa <- readRDS("~/16S_mouse_data/MMUPHIN_SeqTabs/taxtab_MMUPHin.rds")
metadata <- readRDS("~/16S_mouse_data/MMUPHIN_SeqTabs/meta_MMUPHin.rds")
fastafile <- readRDS("~/16S_mouse_data/MMUPHIN_SeqTabs/fastafile_MMUPHin.rds")
alignment <- readRDS("~/16S_mouse_data/MMUPHIN_SeqTabs/alignment_MMUPHin.rds")

OTUmatrix = as.matrix(OTU) #Convert OTU table to matrix
taxamatrix = as.matrix(taxa) #Convert taxtable table to matrix


#Generating a phylogenetic tree
phang.align <- phyDat(as(alignment, "matrix"), type="DNA") #Change sequence alignment output into a phyDat structure
dm <-dist.ml(phang.align) #Create distance matrix
treeNJ <- NJ(dm) #Perform Neighbour joining
fit = pml(treeNJ, data=phang.align) #Internal maximum likelihood
##negative edges length changed to 0!
fitGTR <- update(fit, k=4, inv=0.2)
fitGTR <- optim.pml(fitGTR, model="GTR", optInv=TRUE, optGamma=TRUE, rearrangement = "stochastic", control = pml.control(trace=0))
saveRDS(fitGTR, "~/16S_mouse_data/MMUPHIN_SeqTabs/fitGTR_MMUPHin.rds") #save phylogenetic tree

#Create a phyloseq object
ps <- phyloseq(otu_table(OTUmatrix, taxa_are_rows=TRUE), tax_table(taxamatrix), sample_data(metadata), phy_tree(fitGTR$tree))
ps

saveRDS(ps, "~/16S_mouse_data/MMUPHIN_SeqTabs/ps_MMUPHin.rds") #phyloseq object

#Adding reference sequences to phyloseq object
dna <- Biostrings::DNAStringSet(taxa_names(ps))
names(dna) <- taxa_names(ps)
ps <- merge_phyloseq(ps, dna)
taxa_names(ps) <- paste0("ASV", seq(ntaxa(ps)))
ps


##Adding the lowest taxonomy for an ASV
ps1 <- add_besthit(ps, sep=":")
taxa_names(ps1)[1:2]

##Aggregating ASVs at different taxonomic levels
pseq1 <- aggregate_taxa(ps1, "Phylum") 
pseq2 <- aggregate_taxa(ps1, "Genus") 
pseq3 <- aggregate_taxa(ps1, "Species") 
pseq4 <- aggregate_taxa(ps, "Species") 

##Converting phyloseq object to dataframe
pseq1df <- phyloseq_to_df(pseq1, addtax = T, addtot = F, addmaxrank = F,
                          sorting = NULL)
pseq2df <- phyloseq_to_df(pseq2, addtax = T, addtot = F, addmaxrank = F,
                          sorting = NULL)
pseq3df <- phyloseq_to_df(pseq3, addtax = T, addtot = F, addmaxrank = F,
                          sorting = NULL)
pseq4df <- phyloseq_to_df(pseq4, addtax = T, addtot = F, addmaxrank = F,
                          sorting = NULL)

##Changing files into MMUPHin format
OTU_test <- pseq3df %>%
  unite("Taxa", Domain:Phylum:Class:Order:Family:Genus:Species) #combining the taxonomic levels into 1 taxa column
OTU_test %>% 
  mutate(across(where(is.character), str_trim)) -> OTU_table1
OTU_table1 <- OTU_table1[,-c(1,3)]
OTUtestmatrix <- as.matrix(OTU_table1)
rownames(OTUtestmatrix) <- OTUtestmatrix[,1]
rownames <- OTUtestmatrix[,1]
otumatrix <- OTUtestmatrix[,-1]
class(otumatrix) = "numeric"

write.table(otumatrix, file = "otumatrix_new_MMUPHin.txt", sep="\t")
write.table(metadata, file = "meta_new_MMUPHin.txt", sep="\t")